adding CA to the created certificates

kroko opened the pull request:
This PR addresses these issues

https://github.com/laravel/valet/issues/390  

https://github.com/laravel/valet/issues/296  (import `~/.valet/CA/LaravelValetCASelfSigned.pem` to FF trusted authorities and all valet sites will be trusted)

![screen shot 2018-01-31 at 09 28 41](https://user-images.githubusercontent.com/720976/35610239-2d60e4fc-0669-11e8-8b31-779bd793b3a0.png)

https://github.com/laravel/valet/issues/460 (works when `curl https://mypage.test/ --cacert ~/.valet/CA/LaravelValetCASelfSigned.pem`)

as well as issue I found out while playing with it, namely that `unsecure` previously did not remove old certs from keychain (had tons of them) due to `security delete-certificate` not passing keychain parameter.

As the changes are only in one file it is easy to test them for those who use valet with master branch (`composer global require laravel/valet dev-master`) and are affected by this issue and cannot wait till this gets merged (if it even will). Simply grab `Site.php` head and replace it with file at `~/.composer/vendor/laravel/valet/cli/Valet/Site.php` (backup existing `Site.php` just in case).

---

1)

The logic where the first step is to `untrust` is kept as is, only now it really removes previous cert from Keychain. However manual cleanup needed, read https://github.com/laravel/valet/pull/515#issuecomment-363936795

2)

It checks if _CA_ directory (a new directory, it is created if it does not exist) contains *LaravelValetCASelfSigned.key* && *LaravelValetCASelfSigned.pem*

If yes, then CA has been already created.

3)

If any of those files are not present it creates private key for CA (*LaravelValetCASelfSigned.key*) and generates root certificate (*LaravelValetCASelfSigned.pem*) in _CA_ directory.  

4)

CA is added as trusted to Keychain.

5)

Private key is created for website certificate that will be issued by our new CA.

6)

Create Certificate Signing Request (CSR).  
Certificate will have common name `sitename.tld` and email address `sitename.tld@laravel.valet`  

7)

Issue the certificate for website.

8)

Add issued website certificate and trust it in Keychain.

---

`unsecure()` tries to delete certificate by searching by common name `sitename.tld`.
In addition `unsecure()` also tries to delete certificate by searching by email `sitename.tld@laravel.valet`.

kroko commented:
tested on macOS 10.13.2

drbyte commented on cli/Valet/Site.php:
```diff
@@ -185,13 +185,68 @@ function secure($url)
+        $cName .= $affix;
+
+        $this->cli->runAsUser(sprintf(
+            'openssl req -new -newkey rsa:2048 -days 730 -nodes -x509 -subj "/C=LV/ST=Latvia/O=%s/localityName=Riga/commonName=%s/organizationalUnitName=Developers/emailAddress=noreply@valet.test/" -keyout %s -out %s',
```
Can this be done with less geographic specificity?

kroko replied:
It can be changed. Should it? What do you suggest - some random state (again) in US is a better choice, because...? Personally I see no harm when coder references to himself, unless it does not bite functionality (here it does not). It is my easter egg (come visit), however people with write access in this repo still will do as they see fit.

drbyte replied:
Was just asking a question.

Happy to celebrate the easter egg. :)

Zegnat replied:
Is there some reason the (basically empty) `subj` that is [currently being used for signing requests](https://github.com/laravel/valet/blob/4c243d7d71a0c3fa9cb850d6f90d40ac11bfde91/cli/Valet/Site.php#L239-L242) can’t be used here as well?

```php
        $this->cli->runAsUser(sprintf(
            'openssl req -new -key %s -out %s -subj "/C=/ST=/O=/localityName=/commonName=*.%s/organizationalUnitName=/emailAddress=/" -config %s -passin pass:',
            $keyPath, $csrPath, $url, $confPath
        ));
```

kroko replied:
@Zegnat none that I know of; C,ST,L can be left empty. Considering this line of code raises so many [questions](http://s2.quickmeme.com/img/43/43c6005bf1e35d4742ea9a46341aca9c2883871c37205f84927f95d8d849a907.jpg) I just pushed update where C, ST, L are empty, no time should be spent on this any more.

corneliusio commented:
Unfortunately, this seems to break secured sites for me. Browsers simply refuse to connect. (on macOS macOS 10.13.3)

And just as a rundown on the what I've done to test.
1. Stopped Valet
2. Ran `composer global require laravel/valet dev-master` and made sure that `Site.php` contained the changes.
3. Ran `valet install` just to make sure everything was up-to-date
4. Ran `valet secure` on a handful of sites, all of which then refused connection.

One thing I've found is that any `.crt` files generated are empty, which is likely causing the refused connection.
Is there any debugging information I could provide to help out with this?

kroko commented:
What I did was to try to replicate what @corneliusio is observing was

Updated to macOS 10.13.3 from 10.13.2

Uninstalled Valet

```sh
valet stop
valet uninstall
composer global remove laravel/valet
composer global clearcache
```

Required Valet

```sh
composer global require laravel/valet dev-master
```

Stopped services

```sh
sudo brew services stop --all
```

Replaced contents of `~/.composer/vendor/laravel/valet/cli/Valet/Site.php` with contents of 
https://raw.githubusercontent.com/WARP-LAB/valet/master/cli/Valet/Site.php

Installed Vallet

```
valet install
```

Checked that path is parked with Valet

```sh
valet paths
```

cd'ed into park and created dummy Laravel app

```sh
laravel new ssltest
valet restart
```

Visited http://ssltest.test/, it works.

Secured site

```sh
valet secure ssltest
```

Visited https://ssltest.test/, it works.

---

My bet is on 2nd step listed below.

1. Can you try the method above to reinstall Valet to master branch? You do not need to remove Valet completely (aka `rm -rf ~/.valet`). And while doing so, insert also points 2. and 3. as described below.

2. **especially** `sudo brew services stop --all` (or if you have other services unrelated to valet runing, then explicitly `php71`, `nginx`, `dnsmasq`, see `brew services list`) before you do `valet install`!

3. As it already has broken on you, before `valet install` also do

```sh
rm -rf ~/.valet/Nginx/LaravelValetCASelfSigned*
rm -rf ~/.valet/Certificates/LaravelValetCASelfSigned*
rm -rf ~/.valet/Certificates/site1*
rm -rf ~/.valet/Certificates/site2*
rm -rf ~/.valet/Certificates/siteN*
```
replace `rm -rf ~/.valet/Certificates/site1*` with the name() of the site(s) you tried to secure with this new method. Why you should `rm -rf ~/.valet/Nginx/LaravelValetCASelfSigned*`? Because my guess  is that services were not restarted correctly and that fucked up file handles resulting in a mess.

corneliusio commented:
Welp, I was really hoping that would do the trick. 😞

Ran through this process three times; the second time I completely nuking Valet from my system.—third time I used my second machine to see if it was just something wonky on my system. Made sure all my homebrew services were stopped (and up to date) before installing Valet again.

Unfortunately, I'm still seeing the same issue. Also noticed that after I secure a site, and then unsecure it—I still have to stop/start valet to get a connection again. (though prob not related to this issue in general, might cause other issues)

I'll see if I can't dig in and figure out why I'm getting empty `.crt`'s. Might be helpful if some other folks could give it a test with the above process to confirm if it's just something wonky on my systems.

kroko commented:
@corneliusio please test the one I pushed just now https://raw.githubusercontent.com/WARP-LAB/valet/master/cli/Valet/Site.php (explicit CA serial).

corneliusio commented:
💯 Nailed it man, that did the trick. No failed connection and once I added `LaravelValetCASelfSigned.pem` to my trusted authorities the secured domain loaded right up w/o needing to add an exception.

![](https://media.giphy.com/media/aWRWTF27ilPzy/giphy-downsized.gif)

kroko commented:
1)
The latest commit worked for you?  

2)
And by 
> added `LaravelValetCASelfSigned.pem` to my trusted authorities the secured domain loaded right up w/o needing to add an exception.

you mean added in Firefox? If so, all clear. If you needed it for Safari/Chrome, then something is wrong - they use Keychain, and CA cert as well as webapp cert are trusted _OOB_  (`trustCa()` and `trustCertificate()` methods should take care of it in `valet secure` routine).

kroko commented:
I quite don't get 
> for each TLD I add

`valet domain` changes domain, not adds new domain to listen to.  

for example `valet domain shmest` would change `~/.valet/dnsmasq.conf` from

```
address=/.test/127.0.0.1
listen-address=127.0.0.1
```
to

```
address=/.shmest/127.0.0.1
listen-address=127.0.0.1
```

I did try `valet domain shmest` and things go horribly wrong - `~/.valet/Certificates` gets polluted now as well as `~/.valet/Nginx`. 
But even not looking at code I have a clue whats going on.

1.
After domain change code goes into `~/.valet/Certificates` and looks blindly at all files and assumes that all files are `webappname.crt` and `webappname.key`. But now it's not, as it includes also 

```
LaravelValetCASelfSigned.affix
LaravelValetCASelfSigned.key
LaravelValetCASelfSigned.pem
```

2.
So what it does it assumes there are sites named

```
LaravelValetCASelfSigned.affix
LaravelValetCASelfSigned.pem
LaravelValetCASelfSigned #without .key, as it is stripped in lookup
```

and creates `conf`, `key`, `csr` and `crt` for them.

```
LaravelValetCASelfSigned.affix.conf
LaravelValetCASelfSigned.affix.crt
LaravelValetCASelfSigned.affix.csr
LaravelValetCASelfSigned.affix.key
LaravelValetCASelfSigned.pem.conf
LaravelValetCASelfSigned.pem.crt
LaravelValetCASelfSigned.pem.csr
LaravelValetCASelfSigned.pem.key
```

And totally messes up actual CA as it assumes `LaravelValetCASelfSigned(.key)` is also site

plus it installs those sites in `~/.valet/Nginx`.

I'll look into this issue, but till now, after you've changed TLD, just do

```
rm -rf ~/.valet/Nginx/LaravelValetCASelfSigned*
rm -rf ~/.valet/Nginx/site1*
rm -rf ~/.valet/Nginx/site2*
rm -rf ~/.valet/Nginx/siteN*
rm -rf ~/.valet/Certificates/LaravelValetCASelfSigned*
rm -rf ~/.valet/Certificates/site1*
rm -rf ~/.valet/Certificates/site2*
rm -rf ~/.valet/Certificates/siteN*
valet secure site1
valet secure site2
valet secure siteN
```

corneliusio commented:
Sorry, "add" was just a poor choice of wording. I was simply referring to updating it.

And cool deal, I suspected the issue was something along those lines after taking a closer look at the file names.

Thanks for your work on this man, this is going to be really nice to have setup.

kroko commented:
Before applying the new head, clean `Certificates` and `Nginx` directory from the mess

```
rm -rf ~/.valet/Nginx/LaravelValetCASelfSigned*
rm -rf ~/.valet/Nginx/site1*
rm -rf ~/.valet/Nginx/site2*
rm -rf ~/.valet/Nginx/siteN*
rm -rf ~/.valet/Certificates/LaravelValetCASelfSigned*
rm -rf ~/.valet/Certificates/site1*
rm -rf ~/.valet/Certificates/site2*
rm -rf ~/.valet/Certificates/siteN*
```

After - here you go
https://raw.githubusercontent.com/WARP-LAB/valet/3f11288920e03ee22a4ec30a8ab8b80a8ca890bd/cli/Valet/Site.php
switch domains back and forth, certificates will be autorebuilt. CAs are stored in separate directory `CA`.

corneliusio commented:
That fixed it. Great work!
Not finding any issues now. 🎉

naknode commented:
This is great -- thank you!

kroko commented:
In current Valet release upon TLD change or any other type of _re-securing_ site (i.e., Valet update or just calling `secure` again on already secured site) certificates are kept in Keychain, thus polluting it. They are not deleted, only trust settings are reset to _Use System Defaults_.

Search in _Keychain Access_ app for TLD used by Valet, say `.test` and you get a mess.  

**Latest push should finally fix it, site certificates are truly deleted.**  

**However for it to work** user should clean up Keychain by deleting all duplicate certificates that have been generated by now for each `mywebappname.valettld` except for one (the one with latest Expires time is probably the only one to keep). If not, the logic in `unsecure()` method will still silently fail (thus mess will grow on). Certificate is deleted by CN lookup in `unsecure()`. The reason for the need for the manual cleanup is that if `security delete-certificate` finds more than one match for CN (which it will if there are many certificates for each `mywebappname.valettld`) it will silently fail with `security: "mywebappname.valettld" is ambiguous, matches more than one certificate`. Deletion by SHA-1 hash value in `unsecure()` is... no.

**And I got rid of affix logic.**

corneliusio commented:
🤔 I'm curious, I'm kind of in the dark here, why is deleting via SHA-1 hash not an option?

I was toying around with a script like this:
```
security find-certificate -a -Z -c "mywebappname.valettld" | awk '/SHA-1/{system("sudo security delete-certificate -Z "$NF" -t /Library/Keychains/System.keychain")}'
```
I'm able to get the correct hashes without any issue, but when I loop through to delete them I get `SecTrustSettingsRemoveTrustSettings (user): No Trust Settings were found.`

kroko commented:
Get rid of `-t` (`-t  Also delete user trust settings for this certificate`).  
By leaving it it will untrust (more specifically - set to system defaults) the certificates, but not delete them from keychain (IMHO [man](x-man-page://security) is misleading). It is actually one of things this PR changes.

[Script you mention](https://unix.stackexchange.com/a/227014) will work as advertised.

>  Deletion by SHA-1 hash value in unsecure() is... no.

Man tells you that `-c "mywebappname.valettld"` will 

> Match on name when searching

but it holds back that it will be substring match. In other words it will match (and delete) all`"*mywebappname.valettld*"` and I find this dangerous, because who knows what kind of TLDs people are using for their Valet sites (and how much are they susceptible of deleting also totally different certificates).

However I just had an idea that we could use deletion by SHA-1, just using `security find-certificate -e`. All site certificates would have email `mywebappname.valettld@valet.test`. To implement that would take 2 mins top  - CSR creation phase needs few character addition

```
'.../emailAddress=/...'
```

to

```
'.../emailAddress=%s@valet.test/...', $url
```

and final site certificate will hold `sitename.valettld@valet.test` that we can match against and the danger of deleting some non-Valet related certificate is close to zero.

corneliusio commented:
> but it holds back that it will be substring match.

That's good info to have. 😅

I do really like the sound of tagging all the certificates with a valet email. Def makes tracking them down and deleting them afterward far more straightforward.

kroko commented:
Done, untrust (`unsecure()`) will also delete by cert email field. Deletion by CN matching is kept though as it still applies to old certificates that have no `sitename.tld@laravel.valet` email field.

I had a strong feeling that `unsecure()` should not try to delete certificate only if

```php
if ($this->files->exists($this->certificatesPath().'/'.$url.'.crt')) {
  // then delete
}
```

but always, so that is also changed. if somebody nukes `~/.valet` directory, there is no cert file in fs, but keychain still holds it. On first `secure()` (which calls also `unsecure()`) there should be call to Keychain.

patrickomeara commented:
This works great, thank you so much @kroko 

For anyone looking, you can use `command + shift + .` to display the `.composer` folder when adding the CA to FF. Figuring that out was the most difficult part of this PR.

kroko commented:
@patrickomeara or `cmd+shift+g` to go into hidden directories, or `defaults write com.apple.Finder AppleShowAllFiles true && killall Finder` to show all hidden entities by default in Finder.

mattstauffer commented:
@kroko This looks like incredible work here. Is it ready to go or are y'all still working through it?

Thanks!

kroko commented:
IMHO it is master ready.

* Separate dir to store CA.

* Tests have been done for secure and unsecure.

* Tests have been done for `valet domain` (thus reissuing all certs to different TLD), thanks @corneliusio for spotting this.

* Tests have been done for deleting certificates from keychain. Note that release implementation does not delete certificates from keychain in all cases (`secure`, `unsecure`, `install`, `domain`) where untrust (`unsecure()`)  is called, meaning that most probably people out there have Keychains polluted with certificates. My addition solves it from this point onward. However it is to risky to solve previous mess (users should just go into Keychain Access app, search for their valet certificates by `.valettld` (`.dev`, `.test`). Technically that could be done if deleting certificates by SHA-1 was done on finding certificates using CN, however it might delete valet unrelated certificates (see https://github.com/laravel/valet/pull/515#issuecomment-364182784 ). Thus I do not recommend it. Certificates are deleted by CN (and if more matches are found, it silently fails), and certificates are also deleted by SHA-1 list that was retreived by finding certificates using email field (which is `sitename.tld@laravel.valet` from now on).

* I've tested it on Safari, Chrome (uses Keychain and works OOB) and Firefox.

* No tests have been done on non macOS, so people showing off their devwork in outbounds are not covered, but as far as I know ngrok that you use does not do TLS anyways  

/offtopic And thus as a side note, why force 301 when securing site, secured site could still serve unsecure version at 80. It is simple to implement at `/cli/stubs/secure.valet.conf` - just replace server block serving 80

```
server {
    listen 80;
    server_name VALET_SITE www.VALET_SITE *.VALET_SITE;
    return 301 https://$host$request_uri;
}
```   

to contents of `/cli/stubs/valet.conf`

ReckeDJ commented:
This is great, this PR makes it possible for me to connect a local running API (using valet) to an local iOS simulator with an secured connection. Just import the CA certificate on the simulator, enable full trust and you're good to go! Thanks a lot for creating this PR @kroko !

ajoy39 commented:
Can we get this merged in anytime soon? Blocking Self Signed certs without the option to bypass them is now default behavior in the latest stable versions of Firefox and Chrome.

Kcharle commented:
+1 to get this merged!

tamarazuk commented:
Would love to see this merged as well. I've been running the patch for a number of weeks without issue.

kroko commented:
any updates if/when this will be merged? is TLS treated with angst and nobody wants to take responsibility by merging it?

there have been [5 releases](https://github.com/laravel/valet/releases)! since  https://github.com/laravel/valet/pull/515#issuecomment-364471862 by @adamwathan . is this really such nonpriority although TLS is critical for all things HTTP2 / PWA?

corneliusio commented:
Yeah, I'm a bit discouraged that there hasn't been any feedback given on this in lieu of being merged, considering all of @kroko's hard work on it. 😕

Been manually patching this in for a few months now it's still been working great.

drbyte commented:
> Been manually patching this in for a few months now it's still been working great.

Same here.

adamwathan commented:
I'll take everyone's word for it; someone else will complain if there are any problems :)

Thanks @kroko!

drbyte commented:
Thanks @adamwathan !

ajoy39 commented:
Thanks @adamwathan !

w0rd-driven commented:
Could another release get dropped for this specifically? I figure if it's been in private incubation since Feb 9 it should be good for the general public by now.

bramus commented:
@adamwathan Can you make a new official release with this fix included?

Because right now people are disabling SSL verification _(see https://github.com/laravel/valet/issues/453 and https://github.com/laravel/valet/issues/460 for example)_, which might end up into production if one's not careful.

Thanks!

simonhamp commented:
@kroko is there any particular order of things I need to do to get this working on a previous install? I've `unsecure`'d and re-`secure`'d all relevant sites, but still have issues via cURL/Guzzle.

kinosang commented:
I think it's better not to add the site certs into keychain, just add the CA only.

We can use `security verify-cert -k /Library/Keychains/System.keychain -c <cert.pem>` to test if the cert is trusted (via CA trust line) before do `trustCertificate()`.

kroko commented:
@simonhamp 

>  I've unsecure'd and re-secure'd all relevant sites

but did you actually installed my patch from `dev-master`? 😄 

i haven't touched composer / valet for a while but from heart the full reinstall would look sometikng like this

```sh
# stop vallet
valet stop
# uninstall vallet
valet uninstall
# stop all services (beware, if some other than dnsmasq, php and nginx are running for you)
sudo brew services stop --all
# remove valet from composer and clear cache
composer global remove laravel/valet
composer global clearcache
# ditch the whole valet app dir (not your actual valet projects!!!:) )
rm -rf ~/.valet
# reinstall in composer
composer global require laravel/valet dev-master
# run valet isntall for it to be set up
valet install
# cd into your valet projects (park) directory
cd /path/to/my/vallet/dir
# park it
valet park
# check that path has been included
valet paths
# create some test project in park dir, for example "ssltest" and secure it
valet secure ssltest
# visit it and test in browser
open https://ssltest.test/
# try curl as mentioned in my OP
curl https://ssltest.test/ --cacert ~/.valet/CA/LaravelValetCASelfSigned.pem
```

simonhamp commented:
@kroko yeh I was definitely using your merged stuff and I did a full reinstall.

Since updating to the latest release (`v2.1.1`) running the adjusted version of that final command appears to work:

```
curl https://ssltest.test/ --cacert ~/.config/valet/CA/LaravelValetCASelfSigned.pem
```

simensen commented:
Is this still the best way to use curl w/ local secure valet hosted project?

drbyte commented:
Yes, but there's a shorter way:

Create your own `~/.curlrc` file:
```sh
cacert=/Users/YOURNAME/.config/valet/CA/LaravelValetCASelfSigned.pem
```

... and then you can use curl to hit local valet sites without having to specify the certificate param


(note: .curlrc files can't understand the `~` home-directory shortcut, so you must specify the full path)

patrickomeara commented:
For anyone looking to use this with fopen (eg spatie's medialibrary) you can add this to your `php.ini`

```ini
[openssl]
openssl.cafile="${HOME}/.config/valet/CA/LaravelValetCASelfSigned.pem"
```

jamieburchell commented:
Agree with @kinosang, if Valet is now using a custom or generated root CA to sign all individual site certificates, it's not necessary to add each site's certificate in to the keychain

simonhamp commented:
As @drbyte suggested, a `~/.curlrc` file will help when using curl from the command line.

As @patrickomeara notes, you will need to add something similar to the appropriate `php.ini` (`/usr/local/etc/php/{version}/php.ini`) in order for this to apply to curl in the context of PHP.

In my case, I wasn't able to use interpolation of `${HOME}` for some reason, so I had to use the absolute path to the CA pem file.

Now everything is working how I'd like it to.

simonhamp commented:
Well... I thought it was working. This unfortunately broke the ability to download dependencies via Composer, so I've actually had to do [explicitly append the Valet CA certificate to the OpenSSL root CA](https://github.com/laravel/valet/issues/712#issuecomment-474494455).

Pezhvak commented:
> Well... I thought it was working. This unfortunately broke the ability to download dependencies via Composer, so I've actually had to explicitly append the Valet CA certificate to the OpenSSL root CA bundle:
> 
> ```
> cp /usr/local/etc/openssl/cert.pem /usr/local/etc/openssl/cert.pem.bak && cat ~/.config/valet/CA/LaravelValetCASelfSigned.pem >> /usr/local/etc/openssl/cert.pem
> ```
> 
> This makes a backup of the `cert.pem` (the Root CA file used by curl if you've installed PHP via Homebrew) and then appends the Valet CA certificate to the main cert.pem.
> 
> Doing this, I haven't needed any customisations to my `php.ini` or anything special in my `~/.curlrc`

tried your way, what i get now is this:

`cURL error 77: error setting certificate verify locations: CAfile: /usr/local/etc/openssl/cert.pem CApath: /usr/local/etc/openssl/certs (see http://curl.haxx.se/libcurl/c/libcurl-errors.html)`

simonhamp commented:
@Pezhvak I'm not sure that's related. Are you sure in debugging you haven't inadvertently changed some other path to something invalid? (I found this: https://stackoverflow.com/a/30314795/123696)

Pezhvak commented:
@simonhamp it's been long time ago, i don't have time to check it now

simonhamp commented:
I bumped into this again when upgrading to PHP 7.3, which also happened to upgrade my OpenSSL to 1.1.

In this case, my command needs to change structure a little to look at the correct `cert.pem` file:

```bash
cp /usr/local/etc/openssl@1.1/cert.pem /usr/local/etc/openssl@1.1/cert.pem.bak && cat ~/.config/valet/CA/LaravelValetCASelfSigned.pem >> /usr/local/etc/openssl@1.1/cert.pem
```

eithed commented:
Thanks @simonhamp ! In my case the solution worked for CLI

For web requests, as `php.ini` reported `openssl.cafile` with no value, I've had to, additionally:
`cp /usr/local/etc/openssl@1.1/cert.pem /usr/local/etc/php/7.4/cacert.pem`
and set `curl.cainfo = "/usr/local/etc/php/7.4/cacert.pem"` within php.ini

gabrielchavezme commented:
> I bumped into this again when upgrading to PHP 7.3, which also happened to upgrade my OpenSSL to 1.1.
> 
> In this case, my command needs to change structure a little to look at the correct `cert.pem` file:
> 
> ```shell
> cp /usr/local/etc/openssl@1.1/cert.pem /usr/local/etc/openssl@1.1/cert.pem.bak && cat ~/.config/valet/CA/LaravelValetCASelfSigned.pem >> /usr/local/etc/openssl@1.1/cert.pem
> ```

Thanks!!!

benjibee commented:
@simonhamp I'd suggest using `brew --prefix` to determine the proper path. This should work for either x86 or ARM Macs:

```
cp `brew --prefix`/etc/openssl@1.1/cert.pem `brew --prefix`/etc/openssl@1.1/cert.pem.bak && cat ~/.config/valet/CA/LaravelValetCASelfSigned.pem >> `brew --prefix`/etc/openssl@1.1/cert.pem
```
